gwpy/io/utils.py

```diff
@@ -98,6 +100,50 @@ def gopen(name, *args, **kwargs):
     return fobj
 
 
+def with_open(func=None, mode="r", pos=0):
+    """Decorate a function to ensure the chosen argument is an open file
+
+    Parameters
+    ----------
+    func : `callable`
+        the function to decorate
+
+    mode : `str`, optional
+        the mode with which to open files
+
+    pos : `int`, optional
+        which argument to look at
+
+    Examples
+    --------
+    To ensure that the first argument is an open read-only file, just use
+    the decorator without functional parentheses or arguments:
+
+    >>> @with_open
+    >>> def my_func(pathorfile, *args, **kwargs)
+    >>>     ...
+
+    To ensure that the second argument (position 1) is a file open for writing:
+
+    >>> @with_open(mode="r", pos=1)
+    >>> def my_func(stuff, pathorfile, *args, **kwargs)
+    >>>     stuff.write_to(pathorfile, *args, **kwargs)
+    """
+    def _decorator(func):
+        @wraps(func)
+        def wrapped_func(*args, **kwargs):
+            if not isinstance(args[pos], FILE_LIKE):
+                with open(args[pos], mode=mode) as fobj:
+                    args = list(args)
+                    args[pos] = fobj
+                    return func(*args, **kwargs)
+            return func(*args, **kwargs)
+        return wrapped_func
+    if func:
+        return _decorator(func)
+    return _decorator
```

Comment:
Can you add some in-line code comments here? We're not able to follow exactly what is going on.

Reply:
I've added inline comments to explain the logic internal to the decorated function, the outer scaffolding is just the standard structure to define a decorator that can be called with or without functional brackets, i.e. allowing both

```python
@with_open
def thing():
    ...
```

and

```python
@with_open(mode="w")
def thing():
    ...
```